### PR TITLE
Add auto encoder/decoder

### PR DESCRIPTION
**Pull request by alfonsogarciacaro**
This still needs some work but it's looking good :) Please have a try and let me know what you think, the new API and function names are just tentative (though we need a static member for optional arguments). I only implemented it in Fable side for now, but a similar approach should work for .NET too, though we will likely have issues with generic casting (`unbox< Decoder<obj> > int`). We will need to add an untyped interface (like `IDecoder`) or use something like the crates they talked about in latest F# eXchange.

I also took chance to remove some of the JS code (either in its own file or in `Emit` attributes) and replace it with F# code, to make it more maintainable. I hope that's OK.

Besides that, I replaced rollup because I haven't updated the plugin yet for Fable 2. And I had some issues with the `Thoth.Json.Net.fsproj` reference when compiling the tests for Fable so I added an MSBuild condition. But now that I think about it, maybe it's because I used the `allFiles` option for fable-splitter and removing it may just be enough, hmmm 🤔

### CODE REVIEW

**Comment by alfonsogarciacaro** [comment]
Ok, at the end I removed the `allFiles` option. It looks cleaner that way :)

**Comment by MangelMaxime** [comment]
I used rollup because I needed to read the generate JavaScript when creating the library. And webpack doesn't really fit that need :)

Now, that most of the lib is written it should be ok to use webpack instead.

**Comment by MangelMaxime** [comment]
I wonder is the new implementation of `Encode.list` will fix #22 

And I do prefer the new implementation without the JavaScript files. 👍

**Comment by alfonsogarciacaro** [comment]
> I used rollup because I needed to read the generate JavaScript when creating the library. And webpack doesn't really fit that need :)

Indeed. I actually replaced Rollup with fable-splitter for the tests, so the code should still be readable (and also in multiple files).

> I wonder if the new implementation of Encode.list will fix #22

I added a test for nested lists. Seems to work :)

> We should add an option to specify if we want the json to use camelCase or PascalCase properties. 

I'll leave that as an exercise for you and other Thoth.Json contributors ;)

> Also, is the Fable 2 reflection API allowing us to add attributes to record properties?

Nope, there's very [little reflection info made available](https://github.com/fable-compiler/Fable/blob/ac61fce3b4ebe5a6551f6d48a60142bc5953cffe/src/js/fable-core/Reflection.ts#L16-L19) (only the Type fullname and generics, besides field and case info only for records and generics respectively) and attribute info cannot be accessed at runtime (you could use `CompiledName` but I wouldn't recommend it for this). As I see it, reflection should be used by devs which just want a quick a simple serializer. Those who need more fine-tuning can still write their own decoder. Using attributes would defeat a bit Thot's purpose, as serialization metadata would become  tightly coupled with the type.


Anyways, the good news is I could enable automatic (de)serialization for .NET too! Unfortunately I couldn't follow the same approach as I was getting runtime errors when trying to cast values parsed with `Decoder<obj>` to their proper type. So at the end I had to use Newtonsoft.Json converters (as in Fable 1). I don't like it very much, as it's verbose, error-prone, difficult to test and maintain (especially because the code now is much different with Thoth.Json for Fable), but let see how it goes.

**Comment by MangelMaxime** [comment]
> I'll leave that as an exercise for you and other Thoth.Json contributors ;)

I will do it so I can better understand your code :)

 > Anyways, the good news is I could enable automatic (de)serialization for .NET too! Unfortunately I couldn't follow the same approach as I was getting runtime errors when trying to cast values parsed with Decoder<obj> to their proper type. So at the end I had to use Newtonsoft.Json converters (as in Fable 1). I don't like it very much, as it's verbose, error-prone, difficult to test and maintain (especially because the code now is much different with Thoth.Json for Fable), but let see how it goes.

That's unfortunate, we will need to make it clear that auto mode on .Net side will not validate the data. 

Is this runtime error limitation due to how .Net works ?

**Comment by alfonsogarciacaro** [comment]
Ah, I didn't think of that! Well, Newtonsoft.Json error reporting is quite good: you get the path and the location (line and column) of the issue in the JSON string. Though it's true we can improve [some of the error messages](https://github.com/MangelMaxime/Thoth/pull/30/files#diff-5a0019811a2f26509efbf51645d866b6R58).

Fable erases generics at runtime, so it's possible to do a cast from `obj list` to `int list` but this fails in .NET. I think it's still possible to create individual decoders with more reflection magic but it would be more work. Maybe we can try some corrupted JSON to see if the error messages are good enough.

**Comment by alfonsogarciacaro** [comment]
@MangelMaxime We probably should add more tests and improve the error messages in the .Net converters, but for now I think this is ready to merge. I tested it with one of my work projects and it's working :tada: so it'll be nice to have an alpha release of Thoth.Json compatible with Fable 2.

**Comment by MangelMaxime** [comment]
@alfonsogarciacaro I removed `static member GenerateDecoder(t: System.Type): Decoder<obj> =` from Thoth.Json.Decode to keep the same API between both Fable & .Net version.

But perhaps we can add the missing one to .Net version. Not sure how to do that however

File: src/Thoth.Json/Decode.fs [hunk 1]
```diff
@@ -543,12 +543,21 @@ let private mixedArray msg (decoders: Decoder<obj>[]) (values: obj[]): Result<ob
             | Error _ -> acc
             | Ok result -> decoder value |> Result.map (fun v -> v::result))
 
-let rec private autoDecodeRecordsAndUnions (t: System.Type): Decoder<obj> =
+let inline private lowerCase (s: string) =
+    s |> Seq.mapi (fun i c -> match i with | 0 -> (System.Char.ToLower c) | _ -> c)  |> System.String.Concat
```
Comment (alfonsogarciacaro) [inline src/Thoth.Json/Decode.fs]: This is traversing through all the chars just to lower the first one, slicing fits here nicely: `(string s.[0]).ToLower() + s.[1..]`

**Comment by alfonsogarciacaro** [comment]
Closing as this has been superseded by #38

**Pull request closed without merging**